### PR TITLE
WT-5649 Refactor WT_REF locking, review all WT_REF.addr reads for locking issues

### PR DESCRIPTION
**Pull request by keithbostic**
@kommiharibabu, @tetsuo-cpp: here are some changes to fix problems with looking at the `WT_REF.addr` fields.

In summary, it's not safe look at `WT_REF.addr` (including the associated timestamps), unless we know they won't change underfoot, and that means we either lock them somehow or ensure that checkpoint & eviction are both ignoring the `WT_REF` when we're looking at them.

I tried to come up with some kind of locking scheme, but failed -- we just touch those fields all over the place, and most of the time it's fine, I believe it's just compaction and checkpoint-sync we have to worry about. I'm a little worried about fast-truncate, too, but I didn't spot any problems.

Let me know if you see an approach that leads to a locking solution, I'm happy to back off from this change if there's a better one!

@kommiharibabu, I ended up reworking `bt_sync.c:__sync_ref_obsolete_check()` in major ways, please look at it carefully.

What I did to find everywhere a `WT_REF.addr` was touched was to change the field name and work my way through all the compiler complaints, convincing myself the usage was safe.

If there are any questions about what I'm doing and why I think it's safe, maybe it's worth the three of us meeting and talking through the change, just let me know.

### CODE REVIEW

File: src/reconcile/rec_write.c [hunk 1]
```diff
@@ -2088,23 +2086,7 @@ __rec_write_wrapup(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_PAGE *page)
...
-                previous_state = ref->state;
-                if (previous_state != WT_REF_LOCKED &&
-                  WT_REF_CAS_STATE(session, ref, previous_state, WT_REF_LOCKED))
-                    break;
-            }
-        ret = __wt_ref_block_free(session, ref);
-        if (!F_ISSET(r, WT_REC_EVICT))
-            WT_REF_SET_STATE(ref, previous_state);
-        WT_RET(ret);
+        WT_RET(__wt_ref_block_free(session, ref));
```
Comment (kommiharibabu) [inline src/reconcile/rec_write.c]: Is the `WT_REF_LOCK` not required here? In the function `__wt_ref_block_free ()` we are just copying the address and freeing it.

File: src/btree/bt_sync.c [hunk 2]
```diff
@@ -179,115 +175,140 @@ __sync_ref_list_pop(WT_SESSION_IMPL *session, WT_REF_LIST *rlp, uint32_t flags)
...
+     * If the page is on-disk and obsolete, mark the page as deleted and also set the parent page as
+     * dirty. This is to ensure the parent is written during the checkpoint and the child page
+     * discarded.
+     */
+    newest_stop_ts = WT_TS_NONE;
+    newest_stop_txn = WT_TXN_NONE;
+    obsolete = false;
+    if (previous_state == WT_REF_DISK) {
+        /* There should be an address, but simply skip any page where we don't find one. */
+        if (__wt_ref_addr_copy(session, ref, &addr)) {
```
Comment (kommiharibabu) [inline src/btree/bt_sync.c]: This is the case of `on-page` address and we are forming the `addr` from cell unpack. Can we mention the same in a comment?

File: src/reconcile/rec_write.c (hunk 1 above)
Comment (keithbostic) [reply src/reconcile/rec_write.c]: This was part of the change, removing the locking that was recently added around the call to `__wt_ref_block_free()`. The rule is any function that reads `WT_REF.addr` is responsible for ensuring that reconciliation (by either checkpoint or eviction), does not happen concurrently with the read. That means that reconciliation should no longer have to lock the WT_REF.addr field.

File: src/btree/bt_sync.c (hunk 2 above)
Comment (keithbostic) [reply src/btree/bt_sync.c]: I don't believe there's any guarantee we're unpacking a cell to acquire the address, that is, the page may be on-disk, but the `WT_REF.addr` structure may be filled out.

For example, if a leaf page was just written and evicted, filling in the `WT_REF.addr`, but the parent page has not yet been itself evicted and read back in, subsequent to the leaf page eviction, we're not going to unpack a cell, we're going to use the in-memory structure.

Please let me know if I'm misunderstanding your comment!

File: src/reconcile/rec_write.c (hunk 1 above)
Comment (kommiharibabu) [reply src/reconcile/rec_write.c]: Thanks for the explanation. Understood.

File: src/btree/bt_sync.c (hunk 2 above)
Comment (kommiharibabu) [reply src/btree/bt_sync.c]: Understood.

File: src/include/btree.i [hunk 3]
```diff
@@ -1070,108 +1085,65 @@ __wt_row_leaf_value(WT_PAGE *page, WT_ROW *rip, WT_ITEM *value)
...
+        return (true);
     }
 
-    __wt_ref_info(session, ref, &addr, &size);
-
-    if (addr_buf != NULL) {
-        if (addr != NULL)
-            memcpy(addr_buf, addr, size);
-        *sizep = size;
+    /* If on-page, the pointer references a cell. */
```
Comment (tetsuo-cpp) [inline src/include/btree.i]: Why does this solve the problem of an `addr` getting written to as you're reading it? Couldn't that happen while you're doing this + `memcpy`ing?

File: src/btree/bt_compact.c [hunk 4]
```diff
@@ -24,13 +25,15 @@ __compact_rewrite(WT_SESSION_IMPL *session, WT_REF *ref, bool *skipp)
 
     bm = S2BT(session)->bm;
 
+    /* If the page is clean, test the original addresses. */
+    if (__wt_page_evict_clean(ref->page))
+        return (__wt_ref_addr_copy(session, ref, &addr) ?
+            bm->compact_page_skip(bm, session, addr.addr, addr.size, skipp) :
+            0);
```
Comment (tetsuo-cpp) [inline src/btree/bt_compact.c]: Does this mean that if we're unable to copy the `addr`, then we don't do anything but signal success?

File: src/include/btree.i [hunk 5]
```diff
@@ -61,6 +66,11 @@ __wt_page_is_empty(WT_PAGE *page)
 static inline bool
 __wt_page_evict_clean(WT_PAGE *page)
 {
+    /*
+     * Be cautious modifying this function: it's reading fields set by checkpoint reconciliation,
```
Comment (tetsuo-cpp) [inline src/include/btree.i]: Is it valid to do this? I understand that we won't have a torn write on any of the things that we're checking. But isn't it possible we read `page_state`, then checkpoint writes to `page_state` and `rec_result`, and then we read `rec_result` (i.e. reading a combination that never existed)?

**Comment by keithbostic** [comment]
@agorrod 
> I'm still a bit uncertain of what exact problem is being solved here - but I agree that having confidence that all contents are static while looking at an addr will make our code safer.

This fixes 3 specific problems (along with lots of restructuring):

* reconciliation no longer has to lock to update the `WT_REF` information (we added that locking recently, but I think that was a mistake, checkpoint/eviction fighting for locks with readers or evicting threads is a bad idea),
* checkpoint-sync's review of HS pages was potentially racing with eviction reconciliation,
* compact was potentially racing with checkpoint/eviction reconciliation.

**Pull request merged**